### ISSUE TITLE
`%a` or Org captures fail when displaying with Org (mu4e-conversation)

### ISSUE THREAD

**Issue by Ambrevar** [djcb/mu#1381]
A bug in mu4e/mu4e-conversation was reported here:

https://www.reddit.com/r/emacs/comments/aqhtvf/clash_between_tree_view_of_mu4econversation_and_a/

Indeed, `org-mu4e-store-link` won't do anything if current mode is not either `mu4e-headers-mode` or `mu4e-view-mode`:
```lisp
(defun org-mu4e-store-link ()
  "Store a link to a mu4e query or message."
  (cond
    ;; storing links to queries
    ((eq major-mode 'mu4e-headers-mode)
      (let* ((query (mu4e-last-query))
	      desc link)
	(org-store-link-props :type "mu4e" :query query)
	(setq
	  desc (concat "mu4e:query:" query)
	  link desc)
	(org-add-link-props :link link :description desc)
	link))
      ;; storing links to messages
    ((eq major-mode 'mu4e-view-mode)
      (let* ((msg  (mu4e-message-at-point))
	     (msgid   (or (plist-get msg :message-id) "<none>"))
	     link)
       (org-store-link-props :type "mu4e" :link link
			     :message-id msgid)
       (setq link (concat "mu4e:msgid:" msgid))
       (org-add-link-props :link link
			   :description (funcall org-mu4e-link-desc-func msg))
       link))))
```

Because this function will always be called from a "mu4e" point of view, I suggest changing the second condition to `t` so that it always runs:

```elisp
(defun org-mu4e-store-link ()
  "Store a link to a mu4e query or message."
  (cond
    ;; storing links to queries
    ((eq major-mode 'mu4e-headers-mode)
      (let* ((query (mu4e-last-query))
	      desc link)
	(org-store-link-props :type "mu4e" :query query)
	(setq
	  desc (concat "mu4e:query:" query)
	  link desc)
	(org-add-link-props :link link :description desc)
	link))
      ;; storing links to messages
    (t
      (let* ((msg  (mu4e-message-at-point))
	     (msgid   (or (plist-get msg :message-id) "<none>"))
	     link)
       (org-store-link-props :type "mu4e" :link link
			     :message-id msgid)
       (setq link (concat "mu4e:msgid:" msgid))
       (org-add-link-props :link link
			   :description (funcall org-mu4e-link-desc-func msg))
       link))))
```

What do you think?

**Comment by techmoerror220** [comment]
With your modification, on my system, the problem persists...

**Comment by Ambrevar** [comment]
@techmoerror220: I can't reproduce.  Could you provide an exact recipe with a minimal configuration?

**Comment by techmoerror220** [comment]
Yes. My setup for using `mu4e`, `mu4e-conversation` and `org` links to email messages is as follows. With it ` org-mu4e-store-link` fails to record the email from which I am calling it unless I disable `mu4e-conversation`'s tree view with the shortcut `V` (or deleting `(setq mu4e-conversation-print-function 'mu4e-conversation-print-tree)`).

``` 
(require 'mu4e)
(require 'org-mu4e)        ;;store org-mode links to messages
(require 'mu4e-contrib)

 (when (require 'org-mu4e nil t)
  (dolist (map (list mu4e-view-mode-map mu4e-headers-mode-map)))
  (setq org-mu4e-link-query-in-headers-mode nil))                 ;;store link to message if in header view, not to header query

(defun ambrevar/mu4e-headers ()
  "Like `mu4e' but show the header view.
Default to unread messages if the header buffer does not already exist."
  (interactive)
  (mu4e~start)
  (if (get-buffer "*mu4e-headers*" )
      (switch-to-buffer "*mu4e-headers*")
    (mu4e-headers-search "flag:unread AND NOT flag:trashed")))

(with-eval-after-load 'mu4e
  ;; mu4e-conversation must be enabled here.
  ;; REVIEW: https://github.com/djcb/mu/issues/1258
  (when (require 'mu4e-conversation nil t)
    (global-mu4e-conversation-mode)
    (add-hook
     'mu4e-conversation-after-send-hook
     (lambda ()
       (let ((mu4e-get-mail-command "offlineimap"))
         (mu4e-update-mail-and-index 'run-in-background))))
    (add-hook 'mu4e-view-mode-hook 'auto-fill-mode)))
(autoload 'ambrevar/mu4e-headers "mu4e")

(setq mu4e-conversation-print-function 'mu4e-conversation-print-tree) ;; make tree view the default

(setq  mu4e-headers-auto-update nil)

;; trying to get org capture work with mu4e-conversation in tree view following Ambrevar's advice. 
(defun org-mu4e-store-link ()
  "Store a link to a mu4e query or message."
  (cond
    ;; storing links to queries
    ((eq major-mode 'mu4e-headers-mode)
      (let* ((query (mu4e-last-query))
	      desc link)
	(org-store-link-props :type "mu4e" :query query)
	(setq
	  desc (concat "mu4e:query:" query)
	  link desc)
	(org-add-link-props :link link :description desc)
	link))
      ;; storing links to messages
    (t
      (let* ((msg  (mu4e-message-at-point))
	     (msgid   (or (plist-get msg :message-id) "<none>"))
	     link)
       (org-store-link-props :type "mu4e" :link link
			     :message-id msgid)
       (setq link (concat "mu4e:msgid:" msgid))
       (org-add-link-props :link link
			   :description (funcall org-mu4e-link-desc-func msg))
       link))))
```

**Comment by djcb** [comment]
I've landed a change in `git` that should help with this... does that solve the problem?

**Comment by Ambrevar** [comment]
@techmoerror220 : My bad, the patch I first suggested applied to `org-old-mu4e`, not `org-mu4e`.  @djcb 's commit should be working though.  Can you confirm?

**Comment by techmoerror220** [comment]
For some reason I need both @djcb's change and @ambrevar's patch. Then org's `capture` works in either tree view or its alternative if and only if I start the capture process from anywhere in the message but the header `Compose new message:` or underneath it, which is fine by me. Thanks to you both!

**Comment by andersjohansson** [comment]
I think this change is problematic.

It seems that for me, this new behaviour breaks `org-store-link` in general. When attempting to store a link in some buffer, `org-store-link` calls each function in `(org-store-link-functions)` and returns a link for the first one that returns non-nil. However, the new behaviour of `org-mu4e-store-link` doesn’t fail gracefully now. If not in a headers buffer, it directly attempts to get a message with `mu4e-message-at-point`, which throws a user-error (in mu4e-warn) if it doesn't work. Instead it should just gracefully return nil to let `org-store-link` continue trying other possibilities.

The simple solution seems to be to add a conditional in `org~mu4e-store-link-message` like this:

    (let ((msg (mu4e-message-at-point t)))
      (when msg)
      ... do the rest ...
      )

The optional argument to mu4e-message-at-point avoids throwing an error and returns nil instead.

**Comment by andersjohansson** [comment]
Just to be informative, this assumption of @Ambrevar is not correct:
> Because this function will always be called from a "mu4e" point of view, I suggest changing the second condition to `t` so that it always runs:

As `org-mu4e-store-link` is the `:store` function of the `mu4e` link type, it will be run by `org-store-link` whenever it tries to store a link.

**Comment by djcb** [comment]
Good point -- I wondered why I made the link version mode-specific, and now I remember why. I have pushed a quick fix.

**Comment by andersjohansson** [comment]
I realized there is a further problem with this. `mu4e-message-at-point` either checks if there is a msg text property at point (for headers buffers) or returns whatever message is stored in `mu4e~view-message`. If I have it right (after grepping for this variable) it gets set when viewing a message, but never really unset so it generally just contains the message last viewed. So, even if I have closed all mu4e buffers, open for example an org mode buffer and call `org-store-link`, I will get a link to the mu4e message I viewed a while ago. Not really what is expected.

I dont’t know if this is a good behaviour or not for `mu4e-message-at-point`.

But perhaps it would be more reasonable to keep the old behaviour of only returning a mu4e-link when in an mu4e-buffer, and provide some mechanism for mu4e-conversation to hook into this check and say, "I am also a mu4e-buffer". I think that letting `org-mu4e-store-link` possibly return things outside of the modes where it belongs is bound to cause problems for `org-store-links`.
